daemon/algod/api/server/v1/handlers/handlers.go

```diff
@@ -419,6 +419,9 @@ func WaitForBlock(ctx lib.ReqContext, w http.ResponseWriter, r *http.Request) {
 	}
 
 	select {
```

Comment:
Is it the case that the deadlock happens when waiting for one of the statements after the select block?
If so, the deadlock can still happen when the 1 Minute timer ticks just before the Shutdown, and Server.Stop kicks in just before the statements following the select block are executed.

Reply:
The issue is not waiting for a 1-2 seconds for the `WaitForBlock` to complete. The issue is that this select statement takes 1 minute to complete. The caller which attempted to shut down the process would try again, causing the Stop method of being called again, and attempting to take the lock a second time. This lock would take a minute before it would get release. Unfortunately, this conflicts with the deadlock detection timeout of 30 seconds.

The solution here need to be two-fold. First, once we started shutting down, abort all pointless blocking operations asap. Second, don't use locking where there is no need to.